src/AppInstallerRepositoryCore/Microsoft/PredefinedInstalledSourceFactory.cpp

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include "pch.h"
+#include "Microsoft/PredefinedInstalledSourceFactory.h"
+#include "Microsoft/SQLiteIndex.h"
+#include "Microsoft/SQLiteIndexSource.h"
+
+using namespace std::string_literals;
+using namespace std::string_view_literals;
+
+namespace AppInstaller::Repository::Microsoft
+{
+    namespace
+    {
+        // Populates the index with the ARP entries from the given root.
+        void PopulateIndexFromARP(SQLiteIndex& index, HKEY rootKey)
+        {
+            UNREFERENCED_PARAMETER(index);
+            UNREFERENCED_PARAMETER(rootKey);
+        }
+
+        // Populates the index with the entries from MSIX.
+        void PopulateIndexFromMSIX(SQLiteIndex& index)
+        {
+            using namespace winrt::Windows::ApplicationModel;
+            using namespace winrt::Windows::Management::Deployment;
+
+            PackageManager packageManager;
+            auto packages = packageManager.FindPackagesForUserWithPackageTypes({}, PackageTypes::Main);
+
+            // Reuse the same manifest object, as we will be setting the same values every time.
+            Manifest::Manifest manifest;
+            // Add one installer for storing the package family name.
+            manifest.Installers.emplace_back();
+            // Every package will have the same tags currently.
+            manifest.Tags = { "msix" };
+
+            // Fields in the index but not populated:
+            //  AppMoniker - Not sure what we would put.
+            //  Channel - We don't know this information here.
+            //  Commands - We could open the manifest and look for these eventually.
+            //  Tags - Not sure what else we could put in here.
+            for (const auto& package : packages)
+            {
+                // System packages are part of the OS, and cannot be managed by the user.
+                // Filter them out as there is no point in showing them in a package manager.
+                auto signatureKind = package.SignatureKind();
+                if (signatureKind == PackageSignatureKind::System)
+                {
+                    continue;
+                }
+
+                auto packageId = package.Id();
+                Utility::NormalizedString familyName = Utility::ConvertToUTF8(packageId.FamilyName());
+
+                manifest.Id = familyName;
+                manifest.Name = Utility::ConvertToUTF8(package.DisplayName());
+
+                if (manifest.Name.empty())
+                {
+                    manifest.Name = Utility::ConvertToUTF8(packageId.Name());
+                }
+
+                std::ostringstream strstr;

```

Comment:
>strstr [](start = 35, length = 6)

strstream?

Reply:
strstr forever!

Then you can do `strstr.str()` to get a str from your str str.

---
In reply to: [507501616](https://github.com/microsoft/winget-cli/pull/598#discussion_r507501616) [](ancestors = 507501616)

Reply:
haha ok 

---
In reply to: [508138280](https://github.com/microsoft/winget-cli/pull/598#discussion_r508138280) [](ancestors = 508138280,507501616)